File: vane-portals/src/main/java/org/oddlama/vane/portals/PortalBlueMapLayer.java
package org.oddlama.vane.portals;

import java.util.UUID;
import org.oddlama.vane.annotation.config.ConfigBoolean;
import org.oddlama.vane.annotation.lang.LangMessage;
import org.oddlama.vane.core.lang.TranslatedMessage;
import org.oddlama.vane.core.module.Context;
import org.oddlama.vane.core.module.ModuleComponent;
import org.oddlama.vane.portals.portal.Portal;

public class PortalBlueMapLayer extends ModuleComponent<Portals> {

    @ConfigBoolean(def = false, desc = "If the marker set should be hidden by default.")
    public boolean configHideByDefault;

    @LangMessage
    public TranslatedMessage langLayerLabel;

    @LangMessage
    public TranslatedMessage langMarkerLabel;

    private PortalBlueMapLayerDelegate delegate = null;

    public PortalBlueMapLayer(final Context<Portals> context) {
        super(context.group("BlueMap", "Enable BlueMap integration to show public portals."));
    }

    public void delayedOnEnable() {
        final var plugin = getModule().getServer().getPluginManager().getPlugin("BlueMap");
        if (plugin == null) {
            return;
        }

        delegate = new PortalBlueMapLayerDelegate(this);
        delegate.onEnable(plugin);
    }

    @Override
    public void onEnable() {
        scheduleNextTick(this::delayedOnEnable);
    }

    @Override
    public void onDisable() {
        if (delegate != null) {
            delegate.onDisable();
            delegate = null;
        }
    }

    public void updateMarker(final Portal portal) {
        if (delegate != null) {
            delegate.updateMarker(portal);
        }
    }

    public void removeMarker(final UUID portalId) {
        if (delegate != null) {
            delegate.removeMarker(portalId);
        }
    }

    public void updateAllMarkers() {
        if (delegate != null) {
            delegate.updateAllMarkers();
        }
    }
}

File: vane-portals/src/main/kotlin/org/oddlama/vane/portals/PortalBlueMapLayer.kt
package org.oddlama.vane.portals

import org.oddlama.vane.annotation.config.ConfigBoolean
import org.oddlama.vane.annotation.lang.LangMessage
import org.oddlama.vane.core.lang.TranslatedMessage
import org.oddlama.vane.core.module.Context
import org.oddlama.vane.core.module.ModuleComponent
import org.oddlama.vane.portals.portal.Portal
import java.util.*

/**
 * Module component that integrates public portal markers with BlueMap.
 *
 * @property configHideByDefault whether the BlueMap marker set is hidden by default.
 * @property langLayerLabel localized label for the marker set.
 * @property langMarkerLabel localized HTML marker label template.
 */
class PortalBlueMapLayer(context: Context<Portals?>) :
    ModuleComponent<Portals?>(context.group("BlueMap", "Enable BlueMap integration to show public portals.")) {
    /** Whether the marker set is hidden by default. */
    @JvmField
    @ConfigBoolean(def = false, desc = "If the marker set should be hidden by default.")
    var configHideByDefault: Boolean = false

    /** Localized marker set label. */
    @JvmField
    @LangMessage
    var langLayerLabel: TranslatedMessage? = null

    /** Localized marker label template for portal markers. */
    @JvmField
    @LangMessage
    var langMarkerLabel: TranslatedMessage? = null

    /** Delegate that performs BlueMap API operations. */
    private var delegate: PortalBlueMapLayerDelegate? = null

    /** Enables BlueMap integration after all plugins have finished startup. */
    fun delayedOnEnable() {
        if (module!!.server.pluginManager.getPlugin("BlueMap") == null) return

        delegate = PortalBlueMapLayerDelegate(this)
        delegate!!.onEnable()
    }

    /** Schedules deferred BlueMap integration setup. */
    public override fun onEnable() {
        scheduleNextTick { delayedOnEnable() }
    }

    /** Tears down BlueMap integration state. */
    public override fun onDisable() {
        delegate?.onDisable()
        delegate = null
    }

    /** Updates or creates the BlueMap marker for [portal]. */
    fun updateMarker(portal: Portal) {
        delegate?.updateMarker(portal)
    }

    /** Removes the BlueMap marker for the given portal id. */
    fun removeMarker(portalId: UUID?) {
        delegate?.removeMarker(portalId)
    }

    /** Rebuilds all portal markers in BlueMap. */
    @Suppress("unused")
    fun updateAllMarkers() {
        delegate?.updateAllMarkers()
    }
}
